File: Software/RK4/raytracer.py
from matplotlib import pyplot as plt
import numpy as np
import scipy.misc as smp

CELESTIAL_SPHERE = 1
HORIZON = 0

SPIN = 0.5
SPIN2 = SPIN**2
R1 = 2.*(1. + np.cos((2./3.)*np.arccos(-SPIN)))
R2 = 2.*(1. + np.cos((2./3.)*np.arccos(+SPIN)))
D = 0.01


def ro(r, theta):
    return np.sqrt(r**2. + SPIN2*np.cos(theta)**2.)


def delta(r, theta):
    return r**2. - 2.*r + SPIN2


def sigma(r, theta):
    return np.sqrt((r**2.+SPIN2)**2.-SPIN2*delta(r, theta)*np.sin(theta)**2.)


def alpha(r, theta):
    return ro(r, theta) * np.sqrt(delta(r, theta)) / sigma(r, theta)


def omega(r, theta):
    return 2. * SPIN * r / (sigma(r, theta)**2.)


def omegaH(r, theta):
    return sigma(r, theta) * np.sin(theta) / ro(r, theta)


# Equations for a null geodesic
def P(r, theta, b, q):
    return r**2. + SPIN2 - SPIN*b


def R(r, theta, b, q):
    return P(r, theta, b, q)**2. - delta(r, theta)*((b-SPIN)**2. + q)


def Z(r, theta, b, q):
    return q - np.cos(theta)**2. * ((b**2./(np.sin(theta)**2.)) - SPIN2)


def b0(r):
    # assert(R1 <= r <= R2)
    return - (r**3. - 3.*r**2. + SPIN2*r + SPIN2) / (SPIN*(r-1.))


def q0(r):
    # assert(R1 <= r <= R2)
    r3 = r**3.
    return - (r3*(r3 - 6.*r**2. + 9.*r - 4.*SPIN2)) / (SPIN2*(r-1.)**2.)

B1 = b0(R2)
B2 = b0(R1)

class Camera:
    def __init__(self, r, theta, phi):
        self.r = r
        self.theta = theta
        self.phi = phi

        Omega = 1. / (SPIN + r**(3./2.))
        self.beta = omegaH(r, theta)*(Omega-omega(r, theta))/alpha(r, theta)


class Ray:
    def __init__(self, x, y):
        self.x = x
        self.y = y

        # self.theta = np.arctan2(y, np.sqrt(D**2 - x**2))
        # self.phi = np.arctan2(x, D)

        self.theta = np.arctan2(np.sqrt(D**2.+x**2.), y)
        self.phi = np.arctan2(x, D)+np.pi


def traceRay(cam, ray):
    # Cartesian components of the unit vector N pointing in the direction of
    # the incoming ray
    Nx = np.sin(ray.theta) * np.cos(ray.phi)
    Ny = np.sin(ray.theta) * np.sin(ray.phi)
    Nz = np.cos(ray.theta)

    # Cartesian components ot the direction of motion of the incoming ray as
    # measured by the FIDO
    den = 1. - cam.beta * Ny
    fac = -np.sqrt(1. - cam.beta**2.)

    ny = -Ny + cam.beta / den
    nx = fac * Nx / den
    nz = fac * Nz / den

    # Spherical components on the FIDO's spherical orthonormal basis
    nr = nx
    ntheta = -nz
    nphi = ny

    # Ray's canonical momenta
    Ef = 1. / (alpha(cam.r, ray.theta) + omega(cam.r, ray.theta)*omegaH(cam.r, ray.theta)*nphi)

    pt = -1.
    pr = Ef * ro(cam.r, ray.theta) * nr / np.sqrt(delta(cam.r, ray.theta))
    ptheta = Ef * ro(cam.r, ray.theta) * ntheta
    pphi = Ef * omegaH(cam.r, ray.theta) * nphi

    # Ray's constants
    b = pphi
    q = b**2. + np.cos(ray.theta)**2. * (b**2.*(np.sin(ray.theta)**2.) - SPIN2)

    # print(B1, b, B2)

    # No radial turning points
    if B1 < b < B2 and q < q0(b):
        if(pr > 0.):
            return HORIZON
        else:
            return CELESTIAL_SPHERE
    # There are two radial turning points
    else:
        # Coefficientes of r^4, r^3, r^2, r^1 and r^0 of R(r)
        coefs = [1.,
                 0.,
                 -q - b**2. + SPIN2,
                 2.*q + 2.*(b**2.) - 4.*SPIN*b + 2.*SPIN2,
                 -SPIN2*q]

        # Get maximum of the real roots of R(r) = 0
        roots = np.roots(coefs)

        rup = 0. if np.iscomplex(roots).all() else np.amax(roots[np.isreal(roots)])

        if cam.r < rup:
            return CELESTIAL_SPHERE
        else:
            return HORIZON

if __name__ == '__main__':
    W = H = 500
    camera = Camera(1000., 0., 0.)
    img = [[traceRay(camera, Ray(x, y)) for x in range(-W, W+1)] for y in range(-H, H+1)]
    plt.imshow(img, cmap='Greys', interpolation='nearest')
    plt.show()

File: Software/RaytracerCPU/raytracer.c
#include <stdio.h>
#include <math.h>
#include <assert.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>
// #include <string.h>

#include "common.c"
#include "solver.c"

#define Pi M_PI
#define SYSTEM_SIZE 5

#define DATA_SIZE 2

void getCanonicalMomenta(Real rayTheta, Real rayPhi, Real* pR, Real* pTheta,
                         Real* pPhi){
     // **************************** SET NORMAL **************************** //
     // Cartesian components of the unit vector N pointing in the direction of
     // the incoming ray
     Real Nx = sin(rayTheta) * cos(rayPhi);
     Real Ny = sin(rayTheta) * sin(rayPhi);
     Real Nz = cos(rayTheta);

     // ********************** SET DIRECTION OF MOTION ********************** //
     // Compute denominator, common to all the cartesian components
     Real den = 1. - __camBeta * Ny;

     // Compute factor common to nx and nz
     Real fac = -sqrt(1. - __camBeta*__camBeta);

     // Compute cartesian coordinates of the direction of motion. See(A.9)
     Real nY = (-Ny + __camBeta) / den;
     Real nX = fac * Nx / den;
     Real nZ = fac * Nz / den;

     // Convert the direction of motion to the FIDO's spherical orthonormal
     // basis. See (A.10)
     Real nR = nX;
     Real nTheta = -nZ;
     Real nPhi = nY;

     // *********************** SET CANONICAL MOMENTA *********************** //
     // Compute energy as measured by the FIDO. See (A.11)
     Real E = 1. / (__alpha + __omega * __pomega * nPhi);

     // Set conserved energy to unity. See (A.11)
     // Real pt = -1;

     // Compute the canonical momenta. See (A.11)
     *pR = E * __ro * nR / sqrt(__delta);
     *pTheta = E * __ro * nTheta;
     *pPhi = E * __pomega * nPhi;
}

void getConservedQuantities(Real pTheta, Real pPhi, Real* b, Real* q){
    // ********************* GET CONSERVED QUANTITIES ********************* //
    // Compute axial angular momentum. See (A.12).
    *b = pPhi;

    // Compute Carter constant. See (A.12).
    Real sinT = sin(__camTheta);
    Real sinT2 = sinT*sinT;

    Real cosT = cos(__camTheta);
    Real cosT2 = cosT*cosT;

    Real pTheta2 = pTheta*pTheta;
    Real b2 = pPhi*pPhi;

    *q = pTheta2 + cosT2*((b2/sinT2) - __a2);
}

void setInitialConditions(Real* globalInitCond,Real* globalConstants,
                          Real imageRows, Real imageCols, Real pixelWidth,
                          Real pixelHeight){
    for(int row = 0; row < imageRows; row++){
        for(int col = 0; col < imageCols; col++){
            // Compute pixel unique identifier for this thread
            int pixel = row*imageCols + col;

            // Compute the position in the global array to store the initial
            // conditions of this ray
            Real* initCond = globalInitCond + pixel*SYSTEM_SIZE;

            // Compute the position in the global array to store the constants of
            // this ray
            Real* constants = globalConstants + pixel*2;

            // Compute pixel position in physical units
            Real x = - (col + 0.5 - imageCols/2) * pixelWidth;
            Real y = (row + 0.5 - imageRows/2) * pixelHeight;

            // Compute direction of the incoming ray in the camera's reference
            // frame
            Real rayPhi = Pi + atan(x / __d);
            Real rayTheta = Pi/2 + atan(y / sqrt(__d*__d + x*x));

            // Compute canonical momenta of the ray and the conserved quantites b
            // and q
            Real pR, pTheta, pPhi, b, q;
            getCanonicalMomenta(rayTheta, rayPhi, &pR, &pTheta, &pPhi);
            getConservedQuantities(pTheta, pPhi, &b, &q);

            // Save ray's initial conditions in the global array
            initCond[0] = __camR;
            initCond[1] = __camTheta;
            initCond[2] = __camPhi;
            initCond[3] = pR;
            initCond[4] = pTheta;

            // Save ray's constants in the global array
            constants[0] = b;
            constants[1] = q;
        }
    }
}

void kernel(Real x0, Real xend, Real* devInitCond, Real h, Real hmax,
            Real* devData, int dataSize, int* devStatus,
            Real resolution, int imageRows, int imageCols){
    for(int row = 0; row < imageRows; row++){
        for(int col = 0; col < imageCols; col++){
            // Compute pixel unique identifier for this thread
            int pixel = row*imageCols + col;

            // Array of status flags: at the output, the (x,y)-th element will be
            // set to SPHERE, HORIZON or disk, showing the final state of the ray.
            int* globalStatus = (int*) devStatus;
            globalStatus += pixel;
            int status = *globalStatus;

            // Integrate the ray only if it's still in the sphere. If it has
            // collided either with the disk or within the horizon, it is not
            // necessary to integrate it anymore.
            if(status == SPHERE){
                // Retrieve the position where the initial conditions this block
                // will work with are.
                // Each block, absolutely identified in the grid by blockId, works
                // with only one initial condition (that has N elements, as N
                // equations are in the system). Then, the position of where these
                // initial conditions are stored in the serialized vector can be
                // computed as blockId * N.
                Real* globalInitCond = (Real*) devInitCond;
                globalInitCond += pixel * SYSTEM_SIZE;

                // Pointer to the additional data array used by computeComponent
                Real* globalData = (Real*) devData;
                globalData += pixel * DATA_SIZE;

                // Local arrays to store the initial conditions and the additional
                // data
                Real initCond[SYSTEM_SIZE], data[DATA_SIZE];

                // Retrieve the data from global to local memory :)
                memcpy(initCond, globalInitCond, sizeof(Real)*SYSTEM_SIZE);
                memcpy(data, globalData, sizeof(Real)*DATA_SIZE);

                // Current time
                Real x = x0;

                // Local variable to know how many iterations spent the solver in
                // the current step.
                int iterations = 0;

                // MAIN ROUTINE. Integrate the ray from x to xend, checking disk
                // collisions on the go with the following algorithm:
                //   -> 0. Check that the ray has not collided with the disk or
                //   with the horizon and that the current time has not exceeded
                //   the final time.
                //   -> 1. Advance the ray a step, calling the main RK45 solver.
                //   -> 2. Test whether the ray has collided with the horizon.
                //          2.1 If the answer to the 2. test is negative: test
                //          whether the current theta has crossed theta = pi/2,
                //          and call bisect in case it did, updating its status
                //          accordingly (set it to DISK if the ray collided with
                //          the horizon).
                //          2.2. If the answer to the 2. test is positive: update
                //          the status of the ray to HORIZON.
                status = SolverRK45(&x, xend, initCond, h, xend - x, data,
                                    &iterations);

                // Update the global status variable with the new computed status
                *globalStatus = status;

                // And, finally, update the current ray state in global memory :)
                memcpy(globalInitCond, initCond, sizeof(Real)*SYSTEM_SIZE);
            } // If status == SPHERE
        }
    }
}

double measureKernelTime(int imageRows, int imageCols){
    Real* initCond = (Real*) malloc(imageCols * imageRows * SYSTEM_SIZE *
                                    sizeof(Real));
    Real* constants = (Real*) malloc(imageCols * imageRows * DATA_SIZE *
                                     sizeof(Real));
    int* status = (int*) malloc(imageCols * imageRows * sizeof(int));

    for (size_t i = 0; i < imageRows; i++) {
        for (size_t j = 0; j < imageCols; j++) {
            status[i*imageCols + j] = SPHERE;
        }
    }

    // Sensor physical size
    Real W = 2;
    Real H = 2;

    Real pixelWidth = W / (Real) imageCols;
    Real pixelHeight = H / (Real) imageRows;

    setInitialConditions(initCond, constants, imageRows, imageCols, pixelWidth,
                         pixelHeight);

    Real x0 = 0.;
    Real xend = -150.;
    Real h = -0.001;
    Real resolution = -1;

    clock_t start, end;

    start = clock();
    kernel(x0, xend, initCond, h, xend-x0, constants, DATA_SIZE, status,
           resolution, imageRows, imageCols);
    end = clock();

    FILE *f = fopen("file.txt", "w");
    if (f == NULL)
    {
        printf("Error opening file!\n");
        exit(1);
    }

    for (size_t i = 0; i < imageRows; i++) {
        for (size_t j = 0; j < imageCols; j++) {
            fprintf(f, "%d,", status[i*imageCols + j]);
        }
        fprintf(f, "\n");
    }

    fclose(f);

    free(status);
    free(constants);
    free(initCond);

    Real timeExec = (end-start)/(double)CLOCKS_PER_SEC;
    return timeExec;
}
